File: java/src/com/cable/CableValue.java
/*
 * CableValue - A property value of a Cable node
 * Copyright (C) 2012 Mark Owen and Christian McCarty
 * 
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */ 

package com.cable;

/**
 * A mixed value type, used to store the value of a property in a CableNode.
 * @author Mark Owen
 */
public class CableValue
{
	/**
	 * All the Mixed types
	 */
	public enum Type
	{
		Bool,
		Numeric,
		String,
	}
	
	/**
	 * Factory for creating a variable with the specified boolean value.
	 * @param value The value to set
	 * @return The newly created variable
	 */
	public static CableValue createBool(boolean value)
	{
		CableValue var = new CableValue();
		var.setBool(value);
		return var;
	}

	/**
	 * Factory for creating a variable with the specified numeric value.
	 * @param value The value to set
	 * @return The newly created variable
	 */
	public static CableValue createNumeric(double value)
	{
		CableValue var = new CableValue();
		var.setNumeric(value);
		return var;		
	}

	/**
	 * Factory for creating a variable with the specified String value.
	 * @param value The value to set
	 * @return The newly created variable
	 */
	public static CableValue createString(String value)
	{
		CableValue var = new CableValue();
		var.setString(value);
		return var;		
	}

	/**
	 * Initialize the variable as a numeric with the value zero.
	 */
	public CableValue()
	{
		mType = Type.Numeric;
		NumericPtr ptr = new NumericPtr();
		ptr.value = 0.0;
		mPtr = ptr;
	}
	
	@Override
	public String toString()
	{
		return mType + ":" + asString();
	}

	/**
	 * Returns the type of this variable.
	 * @return The type
	 */
	public Type getType()
	{
		return mType;
	}
	
	/**
	 * Returns true if the value of this is of type bool.
	 * @return True if type is bool
	 */
	public boolean isBool()
	{
		return mType == Type.Bool;
	}

	/**
	 * Returns true if the value of this is of type numeric.
	 * @return True if type is numeric
	 */
	public boolean isNumeric()
	{
		return mType == Type.Numeric;
	}

	/**
	 * Returns true if the value of this is of type string.
	 * @return True if type is string
	 */
	public boolean isString()
	{
		return mType == Type.String;
	}

	/**
	 * Set this variable with the specified value.
	 * @param value The value to set
	 * @throws Error
	 */
	public void set(CableValue value) throws Error
	{
		switch(value.mType)
		{
			case Bool: setBool(value.asBool()); break;
			case Numeric: setNumeric(value.asNumeric()); break;
			case String: setString(value.asString()); break;
		}
	}

	/**
	 * Sets this variable with the specified boolean value.
	 * @param value The value to set
	 */
	public void setBool(boolean value)
	{
		mType = Type.Bool;
		BoolPtr ptr = new BoolPtr();
		ptr.value = value;
		mPtr = ptr;
	}

	/**
	 * Sets this variable with the specified numeric value.
	 * @param value The value to set
	 */
	public void setNumeric(double value)
	{
		mType = Type.Numeric;
		NumericPtr ptr = new NumericPtr();
		ptr.value = value;
		mPtr = ptr;
	}

	/**
	 * Sets this variable with the specified String value.
	 * @param value The value to set
	 */
	public void setString(String value)
	{
		mType = Type.String;
		mPtr = value;
	}
	
	/**
	 * Returns this value as a boolean.
	 * @return This value as a boolean
	 * @throws Error
	 */
	public boolean asBool()
	{
		switch(mType)
		{
			case Bool: return ((BoolPtr)mPtr).value;
			case Numeric: return ((NumericPtr)mPtr).value != 0 ? true : false;
			case String: return ((String)mPtr).length() > 0 ? true : false;
		}

		return false;
	}
	
	/**
	 * Returns this value as a numeric.
	 * @return This value as a numeric
	 * @throws Error
	 */
	public double asNumeric() throws NumberFormatException
	{
		switch(mType)
		{
			case Bool: return ((BoolPtr)mPtr).value ? 1.0 : 0.0;
			case Numeric: return ((NumericPtr)mPtr).value;
			case String: return Double.parseDouble((String)mPtr);
		}

		return 0;
	}
	
	/**
	 * Returns this value as a String.
	 * @return This value as a String
	 * @throws Error
	 */
	public String asString() throws Error
	{
		switch(mType)
		{
			case Bool: return ((BoolPtr)mPtr).value ? "true" : "false";
			case Numeric: return String.valueOf(((NumericPtr)mPtr).value);
			case String: return (String)mPtr;
		}
		
		return "";
	}

	/**
	 * Bool Reference
	 */
	private static class BoolPtr { public boolean value; }

	/**
	 * Numeric Reference
	 */
	private static class NumericPtr { public double value; }

	/**
	 * Type code
	 */
	private Type mType;

	/**
	 * Pointer to the actual data
	 */
	private Object mPtr;
}

File: python/CableValue.py
"""
CableWriter - Used for writing Cable text from the nodes
Copyright (C) 2012  Mark Owen and Christian McCarty

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.
 """ 

class CableValue:
    """
    All the mixed types wrapped within a single interface.
    As well might be guessed, this entire class is worthless in Python.  The 
    interface has been preserved for code transition purposes only; for all
    intents and purposes simply declare a normal variable and code 
    business as usual.
    """
    class _Type:
        Bool = bool
        Numeric = float
        String = unicode
        

    def _createBool(newvalue):
        """ Factory for creating a variable with the specified boolean newvalue.
        Interface compatibility from Java version of Cable.
        """
        return CableValue(bool(newvalue))
    
    def _createNumeric(newvalue):
        """Factory for creating a variable with the specified numeric newvalue.
        Interface compatibility from Java version of Cable.
        """
        return CableValue(float(newvalue))

    def _createString(newvalue):
        """Factory for creating a variable with the specified String newvalue.
        Interface compatibility from Java version of Cable.
        """
        return CableValue(str(newvalue))
        
    def __init__(self, newvalue=0.0):
        self.value = newvalue
        
    def _toString(self):
        """A String representation of the Type and Value of this object
        """
        return type(self.value) + ':' + self.value
    
    def _getType(self):
        """Returns the type of this variable.
        Interface compatibility from Java version of Cable.
        """
        return type(self.value)
        
    def _isBool(self):
        """Returns true if the value of this is of type bool.
        Interface compatibility from Java version of Cable.
        """
        return isinstance(self.value, bool)
        
    def _isNumeric(self):
        """Returns true if the value of this is of type numeric.
        Interface compatibility from Java version of Cable.
        """
        try:
            float(self.value)
            return True
        except:
            return false
            
    def _isString(self):
        """Returns true if the value of this is of type string.
        Interface compatibility from Java version of Cable.
        """
        return isinstance(self.value, basestring)
    
    def _set(self, newvalue):
        """Sets the internal value either by copying another CableValue's data
        or by setting according to passed value.
        """
        try:
            self.value = newvalue.value
        except:
            self.value = newvalue
    
    def _setString(self, newvalue):
        """Sets this variable with the specified String value.
        Interface compatibility from Java version of Cable.
        """
        self.value = str(newvalue)
    
    def _setNumeric(self, newvalue):
        """Sets this variable with the specified numeric value.
        Interface compatibility from Java version of Cable.
        """
        self.value = float(newvalue)
    
    def _setBool(self, newvalue):
        """Sets this variable with the specified boolean newvalue.
        Interface compatibility from Java version of Cable.
        """
        self.value = bool(newvalue)
    
    def _asString(self):
        """Returns this value as a string.
        Interface compatibility from Java version of Cable.
        """
        return unicode(self.value)
    
    def _asNumeric(self):
        """Returns this value as a numeric.
        Interface compatibility from Java version of Cable.
        """
        try:
            return float(self.value)
        except:
            return 0.0

    def _asBool(self):
        """Returns this value as a boolean.
        Interface compatibility from Java version of Cable.
        """
        return bool(self.value)
            

    
    
    
    
    
    




